### PR TITLE
tx_dist based rate estimation

### PR DESCRIPTION
**Pull request by tdaede**

### CODE REVIEW

File: src/rdo.rs [hunk 1]
```diff
@@ -77,6 +91,86 @@ pub struct RDOPartitionOutput {
...
+        RDO_NUM_BINS - 1
+      } else {
+        bin_idx_tmp
+      };
+      self.rate_counts[q_bin_idx][bs_index][bin_idx] += 1;
+      self.rate_bins[q_bin_idx][bs_index][bin_idx] += rate;
+    }
+  }
+  pub fn print_code(&self) {
+    println!("pub static RDO_RATE_TABLE: [[[u64; RDO_NUM_BINS]; TxSize::TX_SIZES_ALL]; RDO_QUANT_BINS] = [");
```
[the remark this answers is not in the document]
Comment (tdaede) [reply src/rdo.rs]: I don't understand this comment?

**Comment by ycho** [comment]
I expect you will include the bdrate change by this commit? I think that will help us to figure out how this new feature impacts and when to enable it.
Also, some brief of how algorithm works will help us learn new thing as well.

File: src/rdo.rs [hunk 2]
```diff
@@ -77,6 +91,86 @@ pub struct RDOPartitionOutput {
...
+  let bs_index = ts as usize;
+  let q_bin_idx = (qindex as usize)/RDO_QUANT_DIV;
+  let bin_idx_down = ((fast_distortion) / RATE_EST_BIN_SIZE).min((RDO_NUM_BINS - 2) as u64);
+  let bin_idx_up = (bin_idx_down + 1).min((RDO_NUM_BINS - 1) as u64);
+  let x0 = (bin_idx_down * RATE_EST_BIN_SIZE) as i64;
+  let x1 = (bin_idx_up * RATE_EST_BIN_SIZE) as i64;
+  let y0 = RDO_RATE_TABLE[q_bin_idx][bs_index][bin_idx_down as usize] as i64;
+  let y1 = RDO_RATE_TABLE[q_bin_idx][bs_index][bin_idx_up as usize] as i64;
+  let slope = ((y1 - y0) << 8) / (x1 - x0);
+  (y0 + (((fast_distortion as i64 - x0) * slope) >> 8)) as u64
```
Comment (barrbrain) [inline src/rdo.rs]: We admit extrapolating beyond the table range based on the final pair of each row. However, if these are decreasing then the estimate may be negative which is then cast to a very large value. Constraining the estimate to non-negative values resolves the overflows seen in tests.
```suggestion
  (y0 + (((fast_distortion as i64 - x0) * slope) >> 8)).max(0) as u64
```

**Comment by tdaede** [comment]
It's still an enormous hit on video: https://beta.arewecompressedyet.com/?job=master-2019-03-04_132251-7ccee27&job=fast_rdo_3-2019-03-04_132256-c7eb04e

I'm going to try retraining it on objective-1-fast.

**Comment by tdaede** [comment]
Retrained, it's still bad: https://beta.arewecompressedyet.com/?job=fast_rdo_3-2019-03-04_163321-32c0d25&job=master-64cf7a982a69b2461d9db1d5ab184eedb765992b

Quite likely the imprecision at low tx_dist.

**Comment by tdaede** [comment]
OK, I disabled it at all speed levels and added the ability to enable it via speed_test. I think I will push it as is.

**Pull request merged**